iceoryx_examples/waitset/ice_waitset_trigger.cpp

```diff
@@ -0,0 +1,234 @@
+            m_actionTrigger.invalidate();
+        }
+        else if (m_activateTrigger.getUniqueId() == uniqueTriggerId)
+        {
+            m_activateTrigger.invalidate();
+        }
+    }
+
+    static void callOnAction(MyTriggerClass* const triggerClassPtr)
+    {
+        std::cout << "action performed" << std::endl;
+    }
+
+  private:
+    uint64_t m_activationCode = 0U;
+    bool m_hasPerformedAction = false;
+    bool m_isActivated = false;
+
+    iox::popo::TriggerHandle m_actionTrigger;
+    iox::popo::TriggerHandle m_activateTrigger;
+};
+
+iox::cxx::optional<iox::popo::WaitSet> waitset;
+iox::cxx::optional<MyTriggerClass> triggerClass;
+
+constexpr uint64_t ACTIVATE_ID = 0U;
+constexpr uint64_t ACTION_ID = 1U;
+
+void callOnActivate(MyTriggerClass* const triggerClassPtr)
+{
+    std::cout << "activated with code: " << triggerClassPtr->getActivationCode() << std::endl;
+}
+
+// The global event loop. It will create an infinite loop and
+// will work on the incoming events.
+void eventLoop()
+{
+    while (true)
+    {
+        auto triggerStateVector = waitset->wait();
+        for (auto& triggerState : triggerStateVector)
+        {
+            if (triggerState.getTriggerId() == ACTIVATE_ID)
+            {
+                // reset MyTriggerClass instance state
+                triggerState.getOrigin<MyTriggerClass>()->reset(MyTriggerClassEvents::ACTIVATE);
+                // call the callback attached to the trigger
+                triggerState();
+            }
+            else if (triggerState.getTriggerId() == ACTION_ID)
+            {
+                // reset MyTriggerClass instance state
+                triggerState.getOrigin<MyTriggerClass>()->reset(MyTriggerClassEvents::PERFORMED_ACTION);
+                // call the callback attached to the trigger
+                triggerState();
+            }
+        }
+    }
+}
+
+int main()
+{
+    iox::runtime::PoshRuntime::initRuntime("/iox-ex-waitset-trigger");
+
+    // we create a waitset and a triggerClass instance inside of the two
+    // global optional's
+    waitset.emplace();
+    triggerClass.emplace();
+
+    // attach both events to a waitset and assign a callback
+    triggerClass->attachTo(*waitset, MyTriggerClassEvents::ACTIVATE, ACTIVATE_ID, callOnActivate);
+    triggerClass->attachTo(*waitset, MyTriggerClassEvents::PERFORMED_ACTION, ACTION_ID, MyTriggerClass::callOnAction);
+
+    // start the event loop which is handling the events
+    std::thread eventLoopThread(eventLoop);
+
+    // start a thread which will trigger a event every second
+    std::thread triggerThread([&] {
+        int activationCode = 1;
+        while (true)
+        {
+            std::this_thread::sleep_for(std::chrono::seconds(1));
+            triggerClass->activate(activationCode++);
+            std::this_thread::sleep_for(std::chrono::seconds(1));
+            triggerClass->performAction();
+        }
+    });
+
+    triggerThread.join();
+    eventLoopThread.join();
```

Comment:
Unless I am mistaken, this will never join (we will not exit the event loop). Maybe add a shutdown trigger to the example as well, to be able to exit with Ctrl-C.

Reply:
fixed